src/coreclrDebug/provisionalDebugSessionTracker.ts

```diff
@@ -0,0 +1,91 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+
+/**
+ * This tracks a debug session until C# dev kit is loaded (if present) and STOPS tracking any existing
+ * debug session after C# dev kit is loaded.
+ * The idea is to avoid a race condition if a debugging session starts before C# dev kit is initialized,
+ * since the brokered service pipe name will be sent with a 'launch' command.
+ * If the extension loads during an existing session, rather than not initializing any C# dev kit services (such as hot reload),
+ * this sends a custom request to the engine with the brokered service pipe name in order to initialize it.
+ */
+export class ProvisionalDebugSessionTracker {
+    private _session: vscode.DebugSession | undefined;
```

Comment:
>_session [](http://example.com/codeflow?start=12&length=8)

I believe we can have more than one debug session

Reply:
is it okay to clear all sessions at `onDidTerminateDebugSession`?

Reply:
I am not positive, but the easiest way to find out is to either:

Option 1 - multiple projects
1. Make a workspace with two console apps
2. Change them so that they will never exit (ex: `while (true) Thread.Sleep(100)`)
3. Create a new launch.json file
4. Add a debug configuration for both projects. Something like below.
5. Add a compond configuration: https://code.visualstudio.com/Docs/editor/debugging#_compound-launch-configurations

```
    {
            "name": "C#: project-name-here Debug",
            "type": "dotnet",
            "request": "launch",
            "projectPath": "${workspaceFolder}/project-name-here.csproj"
        }
```

Option 2: launch the same project more than once

Disclaimer: I haven't tried this, but seems like it should work.

1. Create the same project as above,
2. Open solution explorer and right click on the project
3. Invoke the debug launch command
4. Repeat again

Reply:
I tested and it seems to work (hot reload works on both apps o:)